### ISSUE TITLE
Triplet loss with classification

### ISSUE THREAD

**Issue by lkmklsmn** [road2stat/deep-learning-recipes#1]
Hi Nan,

really helpful blogpost!

I am trying to extend your code and add classification to the triplet problem. In my case I am trying to learn a classifier for two MNIST digits.

First I am generating triplets of 1s and 2s like this:

```
# Load and prepare MNIST data
mnist <- dataset_mnist()
x_train <- mnist$train$x
y_train <- mnist$train$y
x_test <- mnist$test$x
y_test <- mnist$test$y

dim(x_train) <- c(nrow(x_train), 784)
dim(x_test) <- c(nrow(x_test), 784)
x_train <- x_train / 255
x_test <- x_test / 255

ok <- which(y_train %in% c(1,2))
x_tmp <- x_train[ok,]
y_tmp <- y_train[ok]

ones <- which(y_tmp == 1)
ones <- sample(ones)
twos <- which(y_tmp == 2)
twos <- sample(twos)

tmp_one <- do.call(rbind, lapply(ones[1:5000], function(x){
  anchor <- x
  positive <- sample(setdiff(ones, x), 1)
  negative <- sample(twos, 1)
  c(anchor, positive, negative)
}))

tmp_two <- do.call(rbind, lapply(twos[1:5000], function(x){
  anchor <- x
  positive <- sample(setdiff(twos, x), 1)
  negative <- sample(ones, 1)
  c(anchor, positive, negative)
}))

triplets <- rbind(tmp_one, tmp_two)
triplets <- triplets[sample(1:nrow(triplets)), ]
```

Next, I tried to adapt your implementation to my scenario:

```
# Define keras triplet model ####
loss_identity <- function(y_true, y_pred) k_mean(y_pred - 0 * y_true)

# margin-based triplet loss 
loss_margin_triplet <- function(x) {
  embed_user <- x[[1]]
  embed_item_positive <- x[[2]]
  embed_item_negative <- x[[3]]
  
  loss <- k_maximum(
    0.0,
    k_sum(embed_user * embed_item_negative, axis = -1, keepdims = TRUE) -
      k_sum(embed_user * embed_item_positive, axis = -1, keepdims = TRUE) +
      10.0
  )
  
  loss
}

# input layers for items (positive and negative)
input_item_anchor <- layer_input(shape = c(784), name = "input_item_anchor")
input_item_positive <- layer_input(shape = c(784), name = "input_item_positive")
input_item_negative <- layer_input(shape = c(784), name = "input_item_negative")
  
layer_embed_item <- layer_embedding(
  input_dim = 10000, output_dim = 2, name = "embed_item")

embed_item_anchor <- input_item_anchor %>%
  layer_embed_item() %>%
  layer_flatten()
embed_item_positive <- input_item_positive %>%
  layer_embed_item() %>%
  layer_flatten()
embed_item_negative <- input_item_negative %>%
  layer_embed_item() %>%
  layer_flatten()
  
# margin-based triplet loss is the output
loss <- list(embed_item_anchor, embed_item_positive, embed_item_negative) %>%
  layer_lambda(loss_margin_triplet, output_shape = c(2))
  
# define model inputs/outputs
model <- keras_model(
  inputs = c(input_item_anchor, input_item_positive, input_item_negative),
  outputs = loss
)

model %>% compile(loss = loss_identity, optimizer = optimizer_nadam())
```

However, I am not quite sure how to continue. Why do we use the loss_identity() function in the compilation step? Dont we want the triplet loss? 
Also, how would I add another classification layer to the network?

**Comment by nanxstats** [comment]
Hi @lkmklsmn -thanks for the questions!

The identity loss is only an engineering workaround for Keras to use the custom triplet loss: having this identity loss as a "decoy", we can just minimize the margin-based triplet loss (called in the lambda layer) or any custom loss we want to learn the embeddings. I'm not sure if there are more elegant solutions for this now, but this is probably the quickest one I can think of.

Conceptually, I'm not sure if such (Siamese) networks can be adapted to build a classifier. The prediction result will not be about which number category an image belongs to -- depending on the type of triplets you define, it's probably more about if two images represent the same number or not, and that is not strictly "classification".

For the prediction itself, there's probably no need to add extra layers. Remember our goal here is merely to learn the embeddings. After getting the embeddings, you'll be able to do predictions (on if two images are similar or not): https://github.com/road2stat/deep-learning-recipes/blob/master/triplet-loss-keras/metric-auc.R#L3-L13

I guess CNN would be a much better network architecture for image classification and learning image embeddings. The triplet loss is mostly used for special applications, such as the recommender system or face recognition problems. You can learn more about the face recognition problem and tell the difference between face recognition and image classification from Andrew Ng's lecture here: https://www.youtube.com/watch?v=d2XB5-tuCWU.

**Comment by lkmklsmn** [comment]
Thx for the reply!

I guess training a RF on the learned embedding would be one way to "add" classification to it, no?

However, any testing data will have to come in triplets, correct?
How would I construct those triplets? Take one test sample and compare with two classes (2 class classification) from the training data? Would this be overfitting?

**Comment by nanxstats** [comment]
I think that may depend on the specific problem and essentially, your definition of "classification". Take face recognition as an example, the face database has many classes (different people's registered faces), you trained the model with a triplet loss on this data. The testing input is a new face, and the output is: for each existing face, we predict the probability if the input face is the same face as the existing face (class). This can be considered as "classification" under the context of image recognition.

Only the training data is in the form of triplets. On testing data, for recommender systems, depending on what you have and you want to predict, the testing data can come in as a user, an item, or a user-item pair (or generally speaking, a user-item pair eventually). For face recognition, the testing data is usually a face image (as the "user", and the "item" would be all the existing faces in the database).

You can construct triplets offline or online, it can be as simple as random sampling, or as complicated as solving a standalone optimization problem. Here are some examples: https://omoindrot.github.io/triplet-loss

Again, if you only have two classes (or even many more classes) of images and their labels, then (I believe) the CNN would be a much more effective way to build a model for classification (also learning the layered representation for images), since it doesn't have the "hard negative mining" issue like the triplet loss models do.
